core/domain/wipeout_service.py

```diff
@@ -112,40 +113,59 @@ def pre_delete_user(user_id):
         4. Create PendingDeletionRequestModel for the user.
 
     Args:
-        user_id: str. The id of the user to be deleted.
+        user_id: str. The id of the user to be deleted. If the user_id
+            corresponds to a profile user then only that profile is deleted.
+            For a full user, all its associated profile users are deleted too.
     """
-    subscribed_exploration_summaries = (
-        exp_fetchers.get_exploration_summaries_subscribed_to(user_id))
-    explorations_to_be_deleted_ids = [
-        exp_summary.id for exp_summary in subscribed_exploration_summaries
-        if exp_summary.is_private() and
-        exp_summary.is_solely_owned_by_user(user_id)]
-    exp_services.delete_explorations(user_id, explorations_to_be_deleted_ids)
-
-    subscribed_collection_summaries = (
-        collection_services.get_collection_summaries_subscribed_to(user_id))
-    collections_to_be_deleted_ids = [
-        col_summary.id for col_summary in subscribed_collection_summaries
-        if col_summary.is_private() and
-        col_summary.is_solely_owned_by_user(user_id)]
-    collection_services.delete_collections(
-        user_id, collections_to_be_deleted_ids)
-
-    # Set all the user's email preferences to False in order to disable all
-    # ordinary emails that could be sent to the users.
-    user_services.update_email_preferences(user_id, False, False, False, False)
-
-    email = user_services.get_user_settings(user_id, strict=True).email
-    user_services.mark_user_for_deletion(user_id)
-
-    save_pending_deletion_request(
-        wipeout_domain.PendingDeletionRequest.create_default(
-            user_id,
-            email,
-            explorations_to_be_deleted_ids,
-            collections_to_be_deleted_ids
+    all_linked_user_ids = [
+        user.user_id for user in
+        user_services.get_all_profiles_auth_details_by_parent_user_id(user_id)
+    ]
+    all_linked_user_ids.append(user_id)
+    for linked_user_id in all_linked_user_ids:
+        subscribed_exploration_summaries = (
```

Comment:
Some of these calls are expensive and will not apply at all for profile users. In such cases, doing all these calls for the profile users is a bit wasteful. I suggest instead that you add guards as needed in the relevant parts of the codebase (e.g. when a user subscribes to an exploration, you verify that they have the appropriate role before doing so, and add backend tests to catch that case), and then here, you only need to do the more expensive checks for the main user rather than in a for loop for N users.

You'll probably want to consult with @vojtechjelinek on this, too -- he wrote this part of the code.

Reply:
Currently the actions assigned to Learner role will not allow them to create these entries (which I have kept in the if condition) i.e in the backend they are protected by a decorator.

@vojtechjelinek can you PTAL and tell whether the current implementation seems fine?

Reply:
Make sure you still add tests though, as described in the first paragraph above.

Reply:
I have already added the tests to check whether the wipeout happening correctly for profiles or not.

Reply:
@sajalasati When you make comments like this, support your claims with evidence. Provide links to the tests.

Reply:
You can find the tests for profiles at the following places in the test file:
- [test_delete_user_for_profile_user_is_successful](https://github.com/oppia/oppia/blob/ec243140e836ba6fb88eeeb8a36ab872c5896b25/core/domain/wipeout_service_test.py#L1583)
- [test_delete_user_for_full_user_and_its_profiles_is_successful](https://github.com/oppia/oppia/blob/ec243140e836ba6fb88eeeb8a36ab872c5896b25/core/domain/wipeout_service_test.py#L1619)
- [test_pre_delete_profile_users_works_correctly](https://github.com/oppia/oppia/blob/ec243140e836ba6fb88eeeb8a36ab872c5896b25/core/domain/wipeout_service_test.py#L196)
- [test_pre_delete_user_for_full_user_also_deletes_all_profiles](https://github.com/oppia/oppia/blob/ec243140e836ba6fb88eeeb8a36ab872c5896b25/core/domain/wipeout_service_test.py#L212)

Also at places like following I have tested that the particular models for profiles do get deleted, even when the corresponding parent user has certain models which don't exist for the profiles.
- [test_delete_user_with_collections_and_explorations_is_successful](https://github.com/oppia/oppia/blob/ec243140e836ba6fb88eeeb8a36ab872c5896b25/core/domain/wipeout_service_test.py#L1711) 
- [test_delete_user_with_collection_and_exploration_is_successful](https://github.com/oppia/oppia/blob/ec243140e836ba6fb88eeeb8a36ab872c5896b25/core/domain/wipeout_service_test.py#L1669)

The rest of the tests are there in the wipeout service test file.

Reply:
I do not like this approach, this way we will have a lot of request to `ndb` where we can be sure there cannot be anything returned for the profile user. 

I think it would make sense to add a flag to the `PendingDeletionRequest` model that would mark the profile users and then we would only do lightweight wipeout (done only on a subsection of the models) for the profile users. You need to map out which fields can contain profile ID ([here is a list of all the user ID fields in our storage](https://docs.google.com/document/d/1uyrVleg8NZEtlKJi7qeboF9aSSDhQ5KBFSMqMn0veDw/edit)) 

The verification then can be run on all the models to be 100 % sure that we didn't miss anything.

Reply:
@vojtechjelinek Okay so you mean I keep an additional attribute in the `PendingDeletionRequest`, say role of the user, and using that we figure out.

- While I think about all the fields, can you tell how could we filter out the classes only valid for profiles using the map - couldn't understand that by looking at the code, it would be helpful if you can guide me to some file using this map (if there exists any).
- What I have understood is that in the code we just take all the models in the particular file (eg. `storage.user.gae_models`), but those classes themselves don't have role related information stored for them. So I think I will have to hardcode required models for profiles in all the functions applying deletion policy for individual models (based on role check). Sounds good or do you have any other suggestion?
(Note: the among the models mentioned in the list, only a handful of them are needed for profiles).

Also @vojtechjelinek do you think a separate PR would be better for implementing all wipeout related requirements for profiles? (if you foresee a lot of "imp" changes happening)

Reply:
You need to look through the code and determine which storage models can contain a profile ID. 

We can either hardcode them or add them as a new method in all the storage model classes that would contain which roles (or the minimal role required) can create these models. If there is only a handful of models I'm fine with hardcoding them.

I'm fine with this being done in a different PR.

Reply:
@vojtechjelinek Done now. PTAL!